### ISSUE TITLE
How do you deploy it on production for a single image?

### ISSUE THREAD

**Issue by deshwalmahesh** [WongKinYiu/ScaledYOLOv4#222]
I got the model weights as `csp-results.pt` after using your repo and results are awesome. 

With `DarkNet` weights in Yolov4, I got weights as `.weights` and I can use [Opencv Contrib](https://pypi.org/project/opencv-contrib-python/) to read and give predictions as:

```
net = cv2.dnn.readNet("./yolov4-obj_best_1_class.weights", "./yolov4-custom_1_class.cfg")
model = cv2.dnn_DetectionModel(net)
model.setInputParams(size=(416, 416),scale=1/255.)
classes, scores, boxes = model.detect(img_array, 0.51, 0.51)

```

My question is, How can I do the same for the current `pytorch_weights.pt` file?

or if not possible, how can I use `Pytorch` only to give Bounding Boxes for a single Image?

**Comment by Abdullah955** [comment]
are you able to solve it ?

**Comment by deshwalmahesh** [comment]
@Abdullah955  Yes, but I am using `Pytorch` only. Also, it requires you to alter the code for deployment. You have to change the code a bit. I'l ask for a pull request for this. I can share the code soon on my private Repo soon.

**Comment by Abdullah955** [comment]
i've modifed the detect.py code 

`
# Loading Character model
weights = 'best_yolov4-p5.pt'
device1 = select_device('0')
model = attempt_load(weights, map_location=device1)  # load FP32 model



def detect(model,
           output,
           immg,
           source= 'test/',
           save_img=True,
           view_img=True,
           save_txt=True,
           img_size=416,
           device='0',
           conf_thres=0.7,
           iou_thres=0.5,
           classes=None,
           agnostic_nms=True,
          augment=True):
    #print("#"*10)
    
    out, source,  view_img, save_txt, imgsz, conf_thres, iou_thres, classes, agnostic_nms = \
        output, source, view_img, save_txt, img_size, conf_thres, iou_thres, classes, agnostic_nms
    
    webcam = source == '0' or source.startswith('rtsp') or source.startswith('http') or source.endswith('.txt')

    # Initialize
    #device = select_device(device)
    if os.path.exists(out):
        shutil.rmtree(out)  # delete output folder
    os.makedirs(out)  # make new output folder
    half = device.type != 'cpu'  # half precision only supported on CUDA

    # Load model
    model = model
    imgsz = check_img_size(imgsz, s=model.stride.max())  # check img_size
    if half:
        model.half()  # to FP16

    # Second-stage classifier
    classify = False
    if classify:
        modelc = load_classifier(name='resnet101', n=2)  # initialize
        modelc.load_state_dict(torch.load('weights/resnet101.pt', map_location=device)['model'])  # load weights
        modelc.to(device).eval()

    # Set Dataloader
    vid_path, vid_writer = None, None
    if webcam:
        view_img = True
        cudnn.benchmark = True  # set True to speed up constant image size inference
        dataset = LoadStreams(source, img_size=imgsz)
    else:
        save_img = True
        #dataset = LoadImages(source, img_size=imgsz)
        
        #img0 = immg[...,::-1]
        img0 = cv2.cvtColor(immg, cv2.COLOR_RGB2BGR)
        img = letterbox(img0, imgsz, 32)[0]
        # Convert
        img = img[:, :, ::-1].transpose(2, 0, 1)  # BGR to RGB, to 3x416x416
        img = np.ascontiguousarray(img)
        dataset = [['', img, img0, None]]

    # Get names and colors
    names = model.module.names if hasattr(model, 'module') else model.names
    #print(names)
    colors = [[random.randint(0, 255) for _ in range(3)] for _ in range(len(names))]

    # Run inference
    t0 = time.time()
    img = torch.zeros((1, 3, imgsz, imgsz), device=device)  # init img
    _ = model(img.half() if half else img) if device.type != 'cpu' else None  # run once
    
    for path, img, im0s, vid_cap in dataset:
        img = torch.from_numpy(img).to(device)
        img = img.half() if half else img.float()  # uint8 to fp16/32
        img /= 255.0  # 0 - 255 to 0.0 - 1.0
        #print(img)
        if img.ndimension() == 3:
            img = img.unsqueeze(0)

        # Inference
        t1 = time_synchronized()
        pred = model(img, augment=augment)[0]
        
        #print(pred)
        #print("SECOOOOOMD")

        # Apply NMS
        
        pred = non_max_suppression(pred, conf_thres, iou_thres, classes=classes, agnostic=agnostic_nms)
        t2 = time_synchronized()
        # Apply Classifier
        if classify:
            pred = apply_classifier(pred, modelc, img, im0s)
            
        # Process detections
        pred_cord = []
        for i, det in enumerate(pred):  # detections per image
            
            if webcam:  # batch_size >= 1
                p, s, im0 = path[i], '%g: ' % i, im0s[i].copy()
            else:
                p, s, im0 = path, '', im0s
            save_path  = str(Path(out) / Path(p).name)
            txt_path = '' # str(Path(out) / Path(p).stem) + ('_%g' % dataset.frame if dataset.mode == 'video' else '')
            s += '%gx%g ' % img.shape[2:]  # print string
            gn = torch.tensor(im0.shape)[[1, 0, 1, 0]]  # normalization gain whwh
            
            if det is not None and len(det):
                # Rescale boxes from img_size to im0 size
                det[:, :4] = scale_coords(img.shape[2:], det[:, :4], im0.shape).round()                
                
                # Print results
                for c in det[:, -1].detach().unique():
                    n = (det[:, -1] == c).sum()  # detections per class
                    s += '%g %ss, ' % (n, names[int(c)])  # add to string

                # Write results
                for *xyxy, conf, cls in det:
                    
                    if save_txt:  # Write to file
                        xywh = (xyxy2xywh(torch.tensor(xyxy).view(1, 4)) / gn).view(-1).tolist()  # normalized xywh
                        pred_cord.append(torch.tensor(xyxy).view(1, 4).tolist())
                        with open(txt_path + '.txt', 'a') as f:
                            f.write(('%g ' * 5 + '\n') % (cls, *xywh))  # label format

                    if save_img or view_img:  # Add bbox to image
                        
                        
                        xywh = (xyxy2xywh(torch.tensor(xyxy).view(1, 4)) / gn).view(-1).tolist() 
                        label = '%s' % (names[int(cls)])
                        #print(xyxy[0])
                        #plot_one_box(xyxy, im0, label=label, color=colors[int(cls)], line_thickness=2)

            # Print time (inference + NMS)
            return pred_cord, im0
            print('%sDone. (%.3fs)' % (s, t2 - t1))

            # Stream results
            if view_img:
                cv2.imshow(p, im0)
                if cv2.waitKey(1) == ord('q'):  # q to quit
                    raise StopIteration

            # Save results (image with detections)
            if save_img:
                if dataset.mode == 'images':
                    cv2.imwrite(save_path, im0)
                else:
                    if vid_path != save_path:  # new video
                        vid_path = save_path
                        if isinstance(vid_writer, cv2.VideoWriter):
                            vid_writer.release()  # release previous video writer

                        fourcc = 'mp4v'  # output video codec
                        fps = vid_cap.get(cv2.CAP_PROP_FPS)
                        w = int(vid_cap.get(cv2.CAP_PROP_FRAME_WIDTH))
                        h = int(vid_cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
                        vid_writer = cv2.VideoWriter(save_path, cv2.VideoWriter_fourcc(*fourcc), fps, (w, h))
                    vid_writer.write(im0)

    if save_txt or save_img:
        print('Results saved to %s' % Path(out))
        if platform == 'darwin' and not opt.update:  # MacOS
            os.system('open ' + save_path)

    print('Done. (%.3fs)' % (time.time() - t0))
    `

it need more alter but it do the work for testing and you can run it using

`result = detect(model,
            immg = cv2.imread('yourimage.jpg')
           device=device,
           output = 'test_output',
           img_size=416,)`

**Comment by Abdullah955** [comment]
i've created this repo for easy use

https://github.com/Abdullah955/Scaledyolov_inference
